### PR TITLE
[core] Each hook should initialize its state synchronously

### LINKED ISSUE #2293: Initialize the state synchronously with the values of the props
> We could move the initial state to the hook level, instead of having a centralized place.

I don't have the whole pictures of the impacts this change would have.
But a few ideas :

## Approach 1

```ts
const useGridPageSize = (api: GridApiRef, props: Pick<GridComponentProps, 'pageSize, 'onPageSizeChange'>)  => {
  useGridStateInit(apiRef, state => ({ ...state, pagination: { page: props.page ?? 0 }}));
}

const useGridPage = (api: GridApiRef, props: Pick<GridComponentProps, 'page, 'onPageChange'>)  => {
  useGridStateInit(state => ({ ...state, pagination: { ...state.pagination!, pageSize: props.pageSize }}));
}
```

```ts
const useGridStateInit = (apiRef, callback: (state: Partial<GridState>): Partial<GridState>)) => {
  const isInitialized = React.useRef(false);
  const [, setGridState] = useGridState(apiRef);

  if (!isInitialized.current) {
    setGridState(callback);
  }
} 
```

We would have a synchronous state initialization with the value from the props and the hook being responsible for its own state.
But the typing `Partial<GridState>` is not very precise since its relying on the hook order.

The main drawback here is that  could not use selectors in the render that come from hooks later in the feature chain.
To avoid this we must have a state initializer in the hook file / folder, but not in the hook function so that it can be called before the hooks are called.
___

## Approach 2

Each hook can export a state initializer. All hooks are called in a single hook taking care of the state initialization and then running the hooks themselves..

```ts
// useGridPageSoze.ts
export const useGridPageSoze: GridPlugin<'pageSize', 'onPageSizeChange'> = (api, props) => {

}

useGridPage.initState = (state, props) => ({ ...state, pagination: { page: props.pageSize ?? 100 }})
```

```ts
// useGridPage.ts
export const useGridPage: GridPlugin<'page', 'onPageChange'> = (api, props) => {

}

useGridPage.initState = (state, props) => ({ ...state, pagination: { ...state.pagination!, page: props.page ?? 0 }})
```

```ts
// GridPlugin.ts
export type GridPlugin<PropKeys extends keyof GridComponentProps> = {
  (apiRef: GridApiRef, props: Pick<GridComponentProps, PropKeys>): void;
  initState?: (state: Partial<GridState>, props: Pick<GridComponentProps, PropKeys>) => Partial<GridState>
} 
```

```ts
// useGridPlugins.ts
// The plugin must change from a render to another to respect the rule of hooks
const useGridPlugins = (plugins: GridPlugin[]): void => {
  const isInitialized = React.useRef(false)

  if (!isInitialized.current) {
    isInitialized.current = true
    apiRef.current.state = plugins.reduce(
      (acc, plugin) => plugin?.initState(acc) ?? acc, 
      {} as Partial<GridState>
    ) 
    apiRef.current.publishEvent(GridEvents.initState, state)
  }

  for (const plugin in plugins) {
    plugin(apiRef, props);
  }
} 
```

```ts
const X_GRID_PLUGINS: GridPlugin[] = [useGridPageSize, useGridPage]

// XGrid.tsx
const XGrid = (inProps: XGridProps) => {
  const apiRef = useGridApiRef(inProps.apiRef);
  const props = useThemeProps({ props: inProps, name: 'MuiDataGrid' });

  useGridPlugins(X_GRID_PLUGINS)
}
```

With this approach, the 1st time the hooks are called, they already have an initialized state even though they are responsible for the initialization.

The `initState` could allow state mutation if we want since its only called once before any state usage.
___

## Approach 3

Each hook can export a state initializer. The `useXGridComponents` and `useDataGridComponents` are then responsible or initializing the state by manually calling each of those state initializer.
It avoids building a generic plugin system but is more error prone because the initialization order must be correct in both files.

```ts
// useGridPageSize.ts
export const useGridPage = (api: GridApiRef, props: Pick<GridComponentProps, 'page, 'onPageChange'>) => {
}

export const initPageSizeState = (state, props) => ({ ...state, pagination: { page: props.page ?? 0 }})
```

```ts
// useGridPage.ts
export const useGridPage = (api: GridApiRef, props: Pick<GridComponentProps, 'page, 'onPageChange'>) => {
}

export const initPageState = (state, props) => ({ ...state, pagination: { ...state.pagination!, page: props.page ?? 0 }})
```

```ts
// XGrid.tsx
const XGrid = (inProps: XGridProps) => {
  const apiRef = useGridApiRef(inProps.apiRef);
  const props = useThemeProps({ props: inProps, name: 'MuiDataGrid' });

  const isInitialized = React.useRef(false)

  if (!isInitialized.current) {
    isInitialized.current = true
   
    let state: Partial<GridState> = {};
    state = initPageSizeState(state, props);
    state = initiPageState(state, props);
    apiRef.current.state = state as GriState;
    apiRef.current.publishEvent(GridEvents.initState, state);
  }

  useGridPageSize(apiRef, props);
  useGridPage(apiRef, props);
}
```
___

This is a major change not to take lightly of course.

I think it tightly related to the prop-to-state pattern (https://github.com/mui-org/material-ui-x/issues/1562#issuecomment-860247688) and we may want to first take a deeper look at the approach we want to have for the update part.

In summary, maybe its better to wait to have a cleaner view of the approach we want to take.

### PR DESCRIPTION
**Pull request by flaviendelangle**
Fixes #2293 
Part of #924 

Is currently rebased upon #2723

## Overview

- [x] Each hook is now responsible for its state initialization
- [x] The components should never import selectors from optional hook like `useGridColumnReorder`, `useGridColumnResize` and `useGridInfiniteLoader`.  Instead, the feature hook should inject the props in `props.componentsProps?.XXX`
- [x] The state initialization should be synchronous  except for DOM-related information
- [x] Unify model update behaviors : always check if the new model equals the old one before storing it in the state 

I plan to rework the folder structure in a future PR to put back the pro-only features inside the `data-grid-pro` folder to be able to ban the imports from `data-grid` to `data-grid-pro` with ESLint.

## Breaking changes

Merge `state.visibleRows` and `state.filter` into `state.filter` 

```diff
- state.filter
+ state.filter.filterModel
+ gridFilterModelSelector(state) // preferred method
```

```diff
- state.visibleRows.visibleRowsLookup
+ state.filter.visibleRowsLookup
+ gridVisibleRowsLookupSelector(state) // preferred method
```

## Detailed structure of a feature hook

I did not reorder the elements of every feature hooks because the order does not always cause bugs.
For instance, if you instantiate the api methods after the events listeners, it only cause problems if one of the event listeners is an api method (ex: `useGridApiEventHandler(apiRef, GridEvents.rowsSet, apiRef.current.applyFilters);`).

But we should use the following order as much as possible to avoid any issue.

Concerning the hook order, the following hook should be after `useGridSomeOtherFeature` as it uses its state in the render phase.
If possible it should be after `useGridRows` as it depends on it, but if for some reason you have to put it after, it would work.
The dependance is asynchronous (the `GridEvents.rowsSet` is published through a `useEffect`) so `useGridExample` would already be initialized for the 1st event publication even if its after in the feature list).

```ts
/**
 * @requires useGridSomeOtherFeature (state)
 * @requires useGridRows (event)
 */
export const useGridExample = (apiRef: GridApiRef, props: GridComponentProps) => {
  // 1. LOGGER
  const logger = useGridLogger(apiRef, 'useGridExample');

  // 2. STATE INITIALISATION
  useGridStateInit(apiRef, (state) => ({
    ...state,
    example: {
      exampleModel: props.exampleModel ?? getDefaultExampleModel(),
    },
  }));

  // 3. STATE SELECTORS AND UPDATER
  const [, setGridState, forceUpdate] = useGridState(apiRef);
  const someOtherFeatureSubState = useGridSelector(apiRef, gridSomeOtherFeatureSubSelector);


  // 4. SYNCHRONOUS CONTROL STATE REGISTRATION
  useGridRegisterControlState(apiRef, {
    stateId: 'example',
    propModel: props.exampleModel,
    propOnChange: props.onExampleModelChange,
    stateSelector: gridExampleModelSelector,
    changeEvent: GridEvents.exampleChange,
  });

  // 5. API METHODS
  const setExampleModel = React.useCallback<GridExampleApi['setExampleModel']>(
    (model) => {
      const currentModel = gridExampleModelSelector(apiRef.current.state);

      if (currentModel !== model) {
        logger.debug('Setting example model');
  
        setGridState((state) => ({
          ...state,
          example: {
            ...state.example,
            exampleModel: model,
          },
        }));

        // If we have complex post update logic to run after model update, we must call it here
        // If not, remember to call `forceUpdate` here
        apiRef.current.applyExample();
      }
    },
    [apiRef, setGridState, logger],
  );

  const applyExample = React.useCallback<GridExampleApi['applyExample']>(
    () => { /* Some code to run whenever the model or the rows are updated */ },
    []
  )

  const exampleApi: GridExampleApi = {
    setExampleModel,
    applyExample,
  };

  useGridApiMethod(apiRef, exampleApi, 'GridExampleApi');

  // 6. UPDATE EFFECTS
  React.useEffect(() => {
    if (props.exampleModel != null) {
      apiRef.current.setExampleModel(props.exampleModel);
    }
  }, [apiRef, props.exampleModel]);

  // 7. FIRST RUN COMPLEX STATE INITIALIZATION
  // Some logic for the state initialization requires to much of the hook codebase to be placed inside `useGridStateInit`.
  //  We add a callback here to run it synchronously during the 1st render, but after the initialization of the api methods.
  useFirstRender(() => apiRef.current.applyExample());

  // 8. EVENT CALLBACKS
  const handleColumnHeaderClick = React.useCallback(
    () => { /* Some code to run when the user clicks on a column header */ },
    [],
  );

  useGridApiEventHandler(apiRef, GridEvents.rowsSet, this.apiRef.applyExample);
  useGridApiEventHandler(apiRef, GridEvents.columnHeaderClick, handleColumnHeaderClick);
};
```

### CODE REVIEW

**Comment by flaviendelangle** [comment]
@mui-org/x I think its ready enough to start reviewing

File: packages/grid/_modules_/grid/hooks/features/columnResize/useGridColumnResize.tsx [hunk 1]
```diff
@@ -340,4 +351,23 @@ export const useGridColumnResize = (
...
+      ...props.componentsProps,
+      columnsHeader: {
+        ...props.componentsProps?.columnsHeader,
+        resizeCol: resizingColumnField,
+      },
+    }),
+    [props.componentsProps, resizingColumnField],
+  );
+
+  return React.useMemo(
```
Comment (m4theushw) [inline packages/grid/_modules_/grid/hooks/features/columnResize/useGridColumnResize.tsx:366]: Why to return props if they are not used in `useDataGridProComponent`?

File: packages/grid/_modules_/grid/hooks/features/filter/useGridFilter.ts [hunk 2]
```diff
@@ -11,20 +11,36 @@ import { useGridApiEventHandler } from '../../root/useGridApiEventHandler';
...
+import { useGridRegisterControlState } from '../../utils/useGridRegisterControlState';
+import { useGridStateInit } from '../../utils/useGridStateInit';
+import { useFirstRender } from '../../utils/useFirstRender';
+
+const checkFilterModelValidity = (model: GridFilterModel) => {
+  if (model.items.length > 1) {
+    const hasItemsWithoutIds = model.items.find((item) => item.id == null);
+    if (hasItemsWithoutIds) {
+      throw new Error(
+        "The 'id' field is required on filterModel.items when you use multiple filters.",
```
Comment (m4theushw) [inline packages/grid/_modules_/grid/hooks/features/filter/useGridFilter.ts:33]: Why is it required?

```suggestion
        "MUI: The 'id' field is required on `filterModel.items` when you use multiple filters.",
```

File: packages/grid/_modules_/grid/hooks/features/selection/useGridSelection.ts [hunk 3]
```diff
@@ -39,6 +39,19 @@ export const useGridSelection = (apiRef: GridApiRef, props: GridComponentProps):
     return [props.selectionModel];
   }, [props.selectionModel]);
 
+  useGridStateInit(apiRef, (state) => ({ ...state, selection: propSelectionModel ?? [] }));
+
+  const [, setGridState, forceUpdate] = useGridState(apiRef);
+  const rowsLookup = useGridSelector(apiRef, gridRowsLookupSelector);
+
+  const isStateControlled = useGridRegisterControlState(apiRef, {
```
Comment (m4theushw) [inline packages/grid/_modules_/grid/hooks/features/selection/useGridSelection.ts:47]: This new `useGridRegisterControlState` hook has two concerns: registering the state and checking if the prop is controlled. I would call `updateControlState` as before and check the prop to see if it is controlled. Its second purpose is only used once and here.

File: packages/grid/_modules_/grid/models/gridSlotsComponentsProps.ts [hunk 4]
```diff
@@ -1,10 +1,13 @@
+import type { GridColumnsHeaderProps } from '../components/columnHeaders/GridColumnHeaders';
+
 /**
  * Overrideable components props dynamically passed to the component at rendering.
  */
 export interface GridSlotsComponentsProps {
   checkbox?: any;
   columnMenu?: any;
   columnsPanel?: any;
+  columnsHeader?: GridColumnsHeaderProps;
```
Comment (m4theushw) [inline packages/grid/_modules_/grid/models/gridSlotsComponentsProps.ts:10]: Why did you add a new prop? There's no respective slot for it.

**Review by m4theushw — COMMENTED** [review]
I didn't look through all files. I'll do a deeper check later. This PR could have been divided in two: the synchronous state initialization and a second one with the `state.key` -> `keySelector(state)` migration. It would be easier to review with less noise. 😁

File: packages/grid/_modules_/grid/hooks/features/columnResize/useGridColumnResize.tsx (hunk 1 above)
Comment (flaviendelangle) [reply packages/grid/_modules_/grid/hooks/features/columnResize/useGridColumnResize.tsx:366]: Oh I forgot to use them, thanks !

File: packages/grid/_modules_/grid/models/gridSlotsComponentsProps.ts (hunk 4 above)
Comment (flaviendelangle) [reply packages/grid/_modules_/grid/models/gridSlotsComponentsProps.ts:10]: It's related to 

>  The components should never import selectors from optional hook like useGridColumnReorder, useGridColumnResize and useGridInfiniteLoader. Instead, the feature hook should inject the props in props.componentsProps?.XXX

Comment (m4theushw) [reply packages/grid/_modules_/grid/models/gridSlotsComponentsProps.ts:10]: It's also related to https://github.com/mui-org/material-ui-x/issues/2293#issuecomment-914266340.

The purpose of `componentsProps` was to allow to change the props of inner components without having to recreate the entire grid. However, the `GridColumnsHeader` has no prop for the user to change. I agree to add new options to `GridSlotsComponentsProps` for those components that could be relevant to allow users to change their props, but not for internal components that only we need to change.

To provide an alternative, we can use events to get the current column being dragged/resized. In these two cases, I was already advocating that these keys could be removed from the state and replaced with events. See https://github.com/mui-org/material-ui-x/issues/820#issuecomment-897906608 and https://github.com/mui-org/material-ui-x/pull/2337

For the column resizing, the selector is first called in `GridColumnHeadersItemCollection` but its value is only used in `GridColumnHeaderSeparator`. The tree is: 

```
GridColumnHeadersItemCollection -> GridColumnHeaderItem -> GridColumnHeaderSeparator
```

For column dragging, it doesn't change too much. The selector is called in `GridColumnsHeader`. Then, its value is used in this first component and passed to `GridColumnHeaderItem`. There's one component in the middle that doesn't need it. It's not a complex task to replace with events. The three is:

```
GridColumnsHeader -> GridColumnHeadersItemCollection -> GridColumnHeaderItem
```

What do you think about using events?

Comment (flaviendelangle) [reply packages/grid/_modules_/grid/models/gridSlotsComponentsProps.ts:10]: I think its worth taking a look at the future pro-only features and see if the information we want to pass to the components can easily be implemented via events.
To avoid having to find another mechanism in a few weeks.

But on paper, events are probably a good approach.
As long as we accept that the event enum is on the `x-data-grid` version, even for paid-only events.
Which won't be a problem for bundle size.
And for separation of concern, it's not perfect, but if we wanted to really put all the code related to a paid-only feature inside the pro package, then we would have the same discussions for the localization and I don't think we want to split it.

I'll try an event-based implementation tomorrow to see how it looks.

Comment (m4theushw) [reply packages/grid/_modules_/grid/models/gridSlotsComponentsProps.ts:10]: Another PR only removing the selector and listening to the event is better. There's no need to remove the key from the state right now.

Comment (flaviendelangle) [reply packages/grid/_modules_/grid/models/gridSlotsComponentsProps.ts:10]: To merge this PR, I need to remove the selector from the component because the state is never set on the DataGrid, only on the DataGridPro now that its the hook which initialize its own state.

But I can do it on another PR, it just need to be merged before this one.

Comment (flaviendelangle) [reply packages/grid/_modules_/grid/models/gridSlotsComponentsProps.ts]: See #2716

File: packages/grid/_modules_/grid/hooks/features/selection/useGridSelection.ts (hunk 3 above)
Comment (flaviendelangle) [reply packages/grid/_modules_/grid/hooks/features/selection/useGridSelection.ts:47]: What do you mean by "call updateControlState as before"

I can remove the 2nd purpose to simplify the understanding of this hook.
The important point of this hook is that it registers synchronously the controlled state rather than in an effect.

Comment (m4theushw) [reply packages/grid/_modules_/grid/hooks/features/selection/useGridSelection.ts:47]: I was saying that we could remove `useGridRegisterControlState` and call `apiRef.current.updateControlState` directly, but in a synchronous way. The new hook seems to be only a wrapper.

Comment (flaviendelangle) [reply packages/grid/_modules_/grid/hooks/features/selection/useGridSelection.ts:47]: I thought that it would eventually have more logic inside it, but the PR is ready and there is none so I suppose we can indeed remove it :+1:
